configfs/fakertmr/fakertmr.go

```diff
@@ -64,68 +64,70 @@ func readTdx(entry string, attr string) ([]byte, error) {
 	return os.ReadFile(path.Join(entry, attr))
 }
 
-func writeTdx(entry string, attr string, content []byte, indexMap map[int]bool) error {
-	switch attr {
-	case tsmRtmrDigest:
-		// Check if the content is a valid SHA384 hash.
-		if len(content) != crypto.SHA384.Size() {
-			return syscall.EINVAL
-		}
-		// Check if the entry is initialized.
-		content, err := os.ReadFile(filepath.Join(entry, tsmPathIndex))
-		if err != nil {
-			return err
-		}
-		rtmrIndex, err := strconv.Atoi(string(content))
-		if err != nil {
-			return err
-		}
-		if rtmrIndex != 2 && rtmrIndex != 3 {
+func makeWriteTdx(root string) func(entry string, attr string, content []byte, indexMap map[int]bool) error {
+	return func(entry string, attr string, content []byte, indexMap map[int]bool) error {
+		switch attr {
+		case tsmRtmrDigest:
+			// Check if the content is a valid SHA384 hash.
+			if len(content) != crypto.SHA384.Size() {
+				return syscall.EINVAL
+			}
+			// Check if the entry is initialized.
+			content, err := os.ReadFile(filepath.Join(entry, tsmPathIndex))
+			if err != nil {
+				return err
+			}
+			rtmrIndex, err := strconv.Atoi(string(content))
+			if err != nil {
+				return err
+			}
+			if rtmrIndex != 2 && rtmrIndex != 3 {
+				return os.ErrPermission
+			}
+			oldDigest, err := os.ReadFile(filepath.Join(entry, tsmRtmrDigest))
+			if err != nil {
+				return err
+			}
+			newDigest := sha512.Sum384(append(oldDigest[:], content...))
+			if err := os.WriteFile(filepath.Join(entry, tsmRtmrDigest), newDigest[:], 0666); err != nil {
+				return err
+			}
+		case tsmPathIndex:
+			rtmrIndex, e := strconv.Atoi(string(content))
+			if e != nil {
+				return fmt.Errorf("WriteTdx: %v", e)
+			}
+			if rtmrIndex < 0 || rtmrIndex > 3 {
+				return fmt.Errorf("WriteTdx: invalid rtmr index %d. Index can only be a non-negative number", rtmrIndex)
+			}
+			if indexMap[rtmrIndex] {
+				return syscall.EBUSY
+			}
+			indexMap[rtmrIndex] = true
+			if err := os.WriteFile(filepath.Join(entry, tsmPathIndex), content, 0666); err != nil {
+				return err
+			}
+			var rtmrPcrMaps = map[int]string{
+				0: "1,7\n",
+				1: "2-6\n",
+				2: "8-15\n",
+				3: "\n",
+			}
+			tempTsmPathTcgMap := filepath.Join(root, tsmPathTcgMap)
+			if err := os.WriteFile(tempTsmPathTcgMap, []byte(rtmrPcrMaps[rtmrIndex]), 0400); err != nil {
```

Comment:
Please add a comment that you write this and rename to keep the readonly permission.

Reply:
Thanks for the feedback. Done